RIPS/rip-9997.md

```diff
@@ -0,0 +1,198 @@
+discussions-to: https://ethereum-magicians.org/t/rip-7560-native-account-abstraction/16664
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-09-01
+requires: 7560
+---
+
+
+## Abstract
+
+This proposal provides block builders with a mechanism to isolate validation from execution for
+a set of RIP-7560 transactions.
+By doing so we simplify the task of filling a block gas space with RIP-7560 transactions and prevent potential
+denial of service attacks against block builders.
+
+## Motivation
+
+The `AA_TX_TYPE` transaction type that is defined in RIP-7560 completely separates transaction validity
+on a protocol level from ECDSA signatures or any other protocol-defined rules, instead allowing account's EVM code
+to determine a validity of a transaction.
+
+This validation EVM code, however, may be affected by any observable state change within a block.
+A block builder trying to include a set of `AA_TX_TYPE` transactions may face a challenge of finding a combination
+that fits together without invalidating each other.
+
+The block builder has to execute the entire body of one transaction before picking the next one.
+Additionally, it is not feasible to apply any limitations on the transaction execution, while it is expected that
+the validation code will be constrained by the block builder with a ruleset like ERC-7562.
+
+This document proposes a mechanism for block builder to specify a set of `AA_TX_TYPE` transactions that are guaranteed
+to fit together due to a fact that all of their validation frames are executed together first,
+before their respective validation frames.
+
+## Specification
+
+### Bundle header RIP-7560 transaction subtype
+
+```
+BUNDLE_HEADER_SUBTYPE = x
+```
+
+Instead, we propose to introduce an explicit "counter" transaction subtype.
+
+The payload should be interpreted as:
+
+```
+0x04 || 0x01 || rlp([chainId, transactionCount])
+```
+
+Header transactions have a unique hash calculated as follows:
+
+```
+keccak256(AA_TX_TYPE || 0x01 || rlp(chainId, transactionCount, blockNumber, txIndex))
+```
+
+The `blockNumber` and `txIndex` parameters are added to the hash to achieve unique header transaction IDs.
+
+The header transactions are only used to help execution clients determine how many of the `AA_TX_TYPE` transactions
+belong to each individual bundle.
+
+The block is not valid if a header transaction is located anywhere except before a set of `AA_TX_TYPE` transactions.
+
+Header transactions do not affect blockchain state and do not cost any gas.
+
+### Non-atomic validation and execution transaction subtype
+
+```
+BUNDLE_TRANSACTION_SUBTYPE = x
+```
+
+These transactions are completely identical to regular RIP-7560 transactions with the exceptions of being
+composable into bundles.
+
+It is important for the wallets to explicitly opt into this feature, as some Smart Contract Accounts
```

Comment:
I don't think the RIP defines how they should opt in.  Will their `validateTransaction` see this is a different subtype that it may or may not accept?

Reply:
> I don't think the RIP defines how they should opt in. Will their `validateTransaction` see this is a different subtype that it may or may not accept?

Right, I added an elaboration that they need to accept `BUNDLE_TRANSACTION_SUBTYPE` which is the "opt-in".